### PR TITLE
MM-34437 Preventing infinite trial requests

### PR DESCRIPTION
**Pull request by harshilsharma63**
#### Summary
Preventing infinite trial requests from Admin Console.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-34437

#### Release Note
* API additions— new endpoint `GET` `/trial-license/prev` for fetching last used trial license.
* Enterprise - https://github.com/mattermost/enterprise/pull/978

API docs PR - https://github.com/mattermost/mattermost-api-reference/pull/632

### CODE REVIEW

File: api4/license.go [hunk 1]
```diff
@@ -94,7 +97,27 @@ func addLicense(c *Context, w http.ResponseWriter, r *http.Request) {
...
-	license, appErr := c.App.Srv().SaveLicense(buf.Bytes())
+	licenseBytes := buf.Bytes()
+	license, appErr := utils.LicenseFromBytes(licenseBytes)
+	if appErr != nil {
+		c.Err = appErr
+		return
+	}
+
+	if license.IsTrial() {
+		can, err := c.App.Srv().LicenseManager.CanStartTrial()
```
Comment (hahmadia) [inline api4/license.go:108]:
```suggestion
		canStartTrialLicense, err := c.App.Srv().LicenseManager.CanStartTrial()
```

File: api4/license.go [hunk 2]
```diff
@@ -154,6 +177,17 @@ func requestTrialLicense(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	canStart, err := c.App.Srv().LicenseManager.CanStartTrial()
```
Comment (hahmadia) [inline api4/license.go:180]:
```suggestion
	canStartTrialLicense, err := c.App.Srv().LicenseManager.CanStartTrial()
```

File: app/license.go [hunk 3]
```diff
@@ -38,6 +38,25 @@ func (s *Server) LoadLicense() {
...
 	licenseStr := os.Getenv(LicenseEnv)
 	if licenseStr != "" {
+		license, err := utils.LicenseFromBytes([]byte(licenseStr))
+		if err != nil {
+			mlog.Error("Failed to read license set in environment.", mlog.Err(err))
+			return
+		}
+
+		if license.IsTrial() {
+			can, err := s.LicenseManager.CanStartTrial()
```
Comment (hahmadia) [inline app/license.go:48]:
```suggestion
			canStartTrialLicense, err := s.LicenseManager.CanStartTrial()
```

File: i18n/en.json [hunk 4]
```diff
@@ -2016,6 +2016,14 @@
     "id": "api.license.request-trial.bad-request.terms-not-accepted",
     "translation": "You must accept the Mattermost Software Evaluation Agreement and Privacy Policy to request a license."
   },
+  {
+    "id": "api.license.request-trial.can-start-trial.error",
+    "translation": "Could not check if a trial can be started"
+  },
+  {
+    "id": "api.license.request-trial.can-start-trial.not-allowed",
+    "translation": "TODO Not Allowed"
```
Comment (hahmadia) [inline i18n/en.json:2025]: Just make sure we put the correct text here

File: model/license.go [hunk 5]
```diff
@@ -16,6 +17,10 @@ const (
 	LICENSE_RENEWAL_LINK  = "https://mattermost.com/renew/"
 )
 
+var (
+	trialDuration = (time.Hour * 24 * 30) + (time.Hour * 8)
```
Comment (hahmadia) [inline model/license.go:21]: A little more clear operation plus comment to explain won't hurt
```suggestion
	trialDuration = 30 * (time.Hour * 24) + (time.Hour * 8) // 720 hours (30 days) + 8 hours is trial license duration
```

**Review by hahmadia — CHANGES_REQUESTED** [review]
I like the easy to read and well laid out code. It was a pleasant to review.

My only question is so we are not providing a way to over-ride all this if in some special instance the user wants to use another trial and we've allowed it? 
If I'm a customer and I want a second round of trial cause my first one was short, Mattermost approved, how would that work? Would I just be given a custom license at this point?

**Review by sbishel — APPROVED** [review]
Looks good. Nice suggestions by @hahmadia

File: model/license.go (hunk 5 above)
Comment (harshilsharma63) [reply model/license.go]: Done.

File: i18n/en.json (hunk 4 above)
Comment (harshilsharma63) [reply i18n/en.json]: Done.

File: app/license.go (hunk 3 above)
Comment (harshilsharma63) [reply app/license.go]: Done.

File: api4/license.go (hunk 2 above)
Comment (harshilsharma63) [reply api4/license.go:180]: Done.

File: api4/license.go (hunk 1 above)
Comment (harshilsharma63) [reply api4/license.go]: Done.

**Comment by harshilsharma63** [comment]
> I like the easy to read and well laid out code. It was a pleasant to review.
> 
> My only question is so we are not providing a way to over-ride all this if in some special instance the user wants to use another trial and we've allowed it?
> If I'm a customer and I want a second round of trial cause my first one was short, Mattermost approved, how would that work? Would I just be given a custom license at this point?

@sbishel mentioned the same concern and to avoid having a sales-sanctioned trial license type, we can have sales create trial extensions that are anything more than 30 days (say 30 days and a week). Does that work @hahmadia @sbishel ?

**Comment by hahmadia** [comment]
> > I like the easy to read and well laid out code. It was a pleasant to review.
> > 
> > My only question is so we are not providing a way to over-ride all this if in some special instance the user wants to use another trial and we've allowed it?
> > If I'm a customer and I want a second round of trial cause my first one was short, Mattermost approved, how would that work? Would I just be given a custom license at this point?
> 
> @sbishel mentioned the same concern and to avoid having a sales-sanctioned trial license type, we can have sales create trial extensions that are anything more than 30 days (say 30 days and a week). Does that work @hahmadia @sbishel ?

@harshilsharma63 sounds good to me boss

**Review by hahmadia — APPROVED** [review]
Beautiful elegant code 💯💯

File: api4/license.go (hunk 2 above)
Comment (hahmadia) [reply api4/license.go:180]: I think this one was forgotten@harshilsharma63

File: api4/license.go [hunk 6]
```diff
@@ -154,6 +177,17 @@ func requestTrialLicense(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		c.Err = model.NewAppError("requestTrialLicense", "api.license.request-trial.can-start-trial.error", nil, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if !canStart {
```
Comment (hahmadia) [inline api4/license.go:186]:
```suggestion
	if !canStartTrialLicense {
```

**Review by hahmadia — APPROVED** [review]
Don't forget to fix this one too!

File: api4/license.go (hunk 6 above)
Comment (hahmadia) [reply api4/license.go]: I just applied it so we don't forget

**Review by hahmadia — COMMENTED** [review]
Stub Review so that this PR can be removed from PR review list

File: api4/license_test.go [hunk 7]
```diff
@@ -75,6 +81,85 @@ func TestUploadLicenseFile(t *testing.T) {
...
+			Id: "PPPPPPPPPPPPPPPPPPPPPPPPPP",
+			Features: &model.Features{
+				Users: &userCount,
+			},
+			Customer: &model.Customer{
+				Name: "Test",
+			},
+			IsTrial:   true,
+			StartsAt:  mills + 100,
+			ExpiresAt: mills + 100 + 2534400000,
```
Comment (sbishel) [inline api4/license_test.go:142]: Add comment or define this number, so reader knows what it is?

File: model/license.go [hunk 8]
```diff
@@ -262,6 +273,17 @@ func (l *License) ToJson() string {
 	return string(b)
 }
 
+func (l *License) IsTrialLicense() bool {
+	return l.IsTrial || (l.ExpiresAt-l.StartsAt) <= trialDuration.Milliseconds()
+}
+
+func (l *License) IsSanctionedTrial() bool {
```
Comment (sbishel) [inline model/license.go:280]: Add quick unit test for this function.

**Review by sbishel — APPROVED** [review]
Just a couple minor things, otherwise LGTM.

File: api4/license_test.go [hunk 9]
```diff
@@ -75,6 +81,85 @@ func TestUploadLicenseFile(t *testing.T) {
...
+		license := model.License{
+			Id: "AAAAAAAAAAAAAAAAAAAAAAAAAA",
+			Features: &model.Features{
+				Users: &userCount,
+			},
+			Customer: &model.Customer{
+				Name: "Test",
+			},
+			StartsAt:  mills + 100,
+			ExpiresAt: mills + 100 + 2534400000 + 1,
```
Comment (sbishel) [inline api4/license_test.go:105]: Add comment or define this number, so reader knows what it is?

**Review by sbishel — APPROVED** [review]

File: utils/license.go [hunk 10]
```diff
@@ -21,17 +22,55 @@ import (
 	"github.com/mattermost/mattermost-server/v5/utils/fileutils"
 )
 
+// TODO revert this before merging the PR
```
Comment (harshilsharma63) [inline utils/license.go:25]: Note to self: revert this before merging.

File: app/license.go [hunk 11]
```diff
@@ -20,9 +20,10 @@ import (
 )
 
 const (
-	requestTrialURL           = "https://customers.mattermost.com/api/v1/trials"
-	LicenseEnv                = "MM_LICENSE"
-	LicenseRenewalURL         = "https://customers.mattermost.com/subscribe/renew"
+	requestTrialURL = "https://customers.mattermost.com/api/v1/trials"
+	LicenseEnv      = "MM_LICENSE"
+	// TODO revert this before merging the PR
```
Comment (harshilsharma63) [inline app/license.go:25]: Note to self: revert this before merging.

**Comment by harshilsharma63** [comment]
The one test fails due to the temporary public key added for testing. It will be fine once we revert the key.

File: api4/license_test.go (hunk 7 above)
Comment (harshilsharma63) [reply api4/license_test.go]: Done.

File: api4/license_test.go (hunk 9 above)
Comment (harshilsharma63) [reply api4/license_test.go]: Done.

File: model/license.go (hunk 8 above)
Comment (harshilsharma63) [reply model/license.go:280]: Done. Added missing tests for `IsTrialLicense()` as well.

**Review by hahmadia — APPROVED** [review]
API changes look good to me now. Nice job!

**Review by sbishel — APPROVED** [review]

File: app/license.go [hunk 12]
```diff
@@ -20,9 +20,11 @@ import (
 )
 
 const (
-	requestTrialURL           = "https://customers.mattermost.com/api/v1/trials"
-	LicenseEnv                = "MM_LICENSE"
-	LicenseRenewalURL         = "https://customers.mattermost.com/subscribe/renew"
+	// TODO revert this before merging
+	requestTrialURL = "https://portal.test.cloud.mattermost.com/api/v1/trials"
```
Comment (harshilsharma63) [inline app/license.go:24]: Note to self: revert this before merging.

**Comment by furqanmlk** [comment]
@harshilsharma63 
 30 days trail license still can be uploaded after remove trial license

**Review by furqanmlk — APPROVED** [review]
Thanks @harshilsharma63 
Following Scenarios are passed

<google-sheets-html-origin><style type="text/css"></style>
Case | Expected | Actaul | License Type | Applied With API | Result
-- | -- | -- | -- | -- | --
With Flag 29 days | Should Apply | Is Applied | Trial | Is Applied | Pass
Without Flag 29 days | Should Apply | Is Applied | Paid | Is Applied | Pass
With Flag less than 29 days | Should Apply | Is Applied | Trial | Is Applied | Pass
With Flag 30 days | Should Not Apply | Did Not Apply | Trial | Did Not Apply | Pass
Without Flag 30 days | Should Not Apply | Did Not Apply | Trial | Did Not Apply | Pass
With Flag 31 days | Should Apply | Is Applied | Trial | Is Applied | Pass
With Flag more than 31 days | Should Apply | Is Applied | Trial | Is Applied | Pass
Without Flag more than 31 days | Should Apply | Is Applied | Paid | Is Applied | Pass

</google-sheets-html-origin>

File: utils/license.go (hunk 10 above)
Comment (harshilsharma63) [reply utils/license.go:25]: Done.

File: app/license.go (hunk 12 above)
Comment (harshilsharma63) [reply app/license.go:24]: Done.

File: app/license.go (hunk 11 above)
Comment (harshilsharma63) [reply app/license.go:26]: Done.

**Comment by harshilsharma63** [comment]
@amyblais this can be merged now. I'm not sure the `do-not-merge/release-note-label-needed` is there since I did add the release notes.

**Pull request merged**